core/models.py

```diff
@@ -116,10 +149,23 @@ class PostVote(models.Model):
         (DOWNVOTE, "Down Vote"),
     ]
 
-    user = models.ManyToManyField(User, related_name="post_votes")
-    post = models.ManyToManyField(Post, related_name="post_votes")
-    type = models.CharField(max_length=20, choices=VOTE_CHOICES)
+    user = models.ForeignKey(User, on_delete=models.CASCADE, related_name="post_votes")
+    post = models.ForeignKey(Post, on_delete=models.CASCADE, related_name="post_votes")
+    choice = models.CharField(max_length=20, choices=VOTE_CHOICES)
     created_at = models.DateTimeField(auto_now_add=True)
 
-    def __str__(self) -> str:  # noqa: ANN101, D105
-        return f"@{self.user}: {self.type} for post: {self.post}"
+    class Meta:
+        unique_together: ClassVar[list[str]] = ["post", "user"]
+
+    def __str__(self: "PostVote") -> str:
+        return f"@{self.user}: {self.choice} for post: {self.post}"
+
+    def save(self: "PostVote", *args: int, **kwargs: int) -> None:
+        super().save(*args, **kwargs)
+        post_vote = PostVote.objects.filter(post=self.post)
+        self.post.up_votes = post_vote.filter(choice=PostVote.UPVOTE).count()

```

Comment:
This should be optimized a bit, using "update" method.
1. since vote is either up or down, it could count only this no?
2. instead of running save -> which will actually alter updated at, we need to run update... using this:
https://docs.djangoproject.com/en/5.0/topics/db/queries/#updating-multiple-objects-at-once
```
post.objects.filter(pk=self.post.pk).update(down_votes={some number}
```
This has the following advantages:
- does not call any signals on objects
- does not call save() on objects
- does not update changed_at value (which will get changed when save is called) and which we need to avoid in this scenario
- it simply alters “silently” one column on the object in db